src/Namespace.js

```diff
@@ -0,0 +1,44 @@
+// @flow
+
+import builtinMacros from "./macros";
+import type {MacroMap, MacroExpansion} from "./macros";
+
+export default class Namespace {
+    parent: ?Namespace;
+    macros: MacroMap;
+    //tag: any;
+
+    constructor(parent: ?Namespace, macros: MacroMap = {}) {
+        this.parent = parent;
+        this.macros = macros;
+    }
+
+    getMacro(name: string): ?MacroExpansion {
+        let ns = this;
+        while (ns && !ns.macros.hasOwnProperty(name)) {
+            ns = ns.parent;
+        }
```

Comment:
Wouldn’t this increase the complexity to O(k), where k is number of namespaces(groups)? Though, I’m not sure how much it will affect the performance.

Reply:
I'm a little concerned about performance.  We could do some benchmarking and/or profiling to see how much this affects things.  We could maybe save a little bit of look up time by reusing the namespace from the parent group and only creating a new one if we run into a `\def` statement.  If we also reuse the `builtinNamespace` in the same way then this should have almost zero impact on users who aren't using `\def` b/c each lookup will be O(1).